File: TaskManagementBackend/TaskManagementApi/Services/Projects/ProjectService.cs
using TaskManagementApi.DTOs.Project;
using TaskManagementApi.Mappers;
using TaskManagementApi.Models;
using TaskManagementApi.Repository.Projects;

namespace TaskManagementApi.Services.Projects;

public class ProjectService : IProjectService
{
    private readonly IProjectRepository _projectRepository;

    public ProjectService(IProjectRepository projectRepository)
    {
        _projectRepository = projectRepository;
    }

    public List<ProjectResponse> GetAll()
    {
        return _projectRepository.GetAll();
    }

    public ProjectResponse GetById(int id)
    {
        ProjectResponse? project = _projectRepository.GetById(id);
        if (project == null)
            throw new Exception($"Project with id {id} not found.");

        return project;
    }

    public List<ProjectResponse> GetByUserId(int userId)
    {
        return _projectRepository.GetByUserId(userId);
    }

    public void Add(CreateProjectRequest request, int createdBy)
    {
        // createdBy comes from JWT claims in the controller — not from the client body
        Project project = request.ToEntity(createdBy);
        _projectRepository.Add(project);
    }

    public void Update(int id, CreateProjectRequest request)
    {
        ProjectResponse? existing = _projectRepository.GetById(id);
        if (existing == null)
            throw new Exception($"Project with id {id} not found.");

        Project project = request.ToEntity(existing.CreatedBy);
        project.Id        = id;
        project.CreatedAt = existing.CreatedAt;
        project.UpdatedAt = DateTime.UtcNow;

        _projectRepository.Update(project);
    }

    public void Delete(int id)
    {
        ProjectResponse? project = _projectRepository.GetById(id);
        if (project == null)
            throw new Exception($"Project with id {id} not found.");

        _projectRepository.Delete(id);
    }
}

File: TaskManagementFrontend/task-management-frontend/src/app/core/services/project.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { CreateProjectRequest, ProjectResponse } from '../models/project.model';

@Injectable({ providedIn: 'root' })
export class ProjectService {
  private readonly apiUrl = 'http://localhost:5114/api/projects';

  constructor(private http: HttpClient) {}

  getAll(): Observable<ProjectResponse[]> {
    return this.http.get<ProjectResponse[]>(this.apiUrl);
  }

  getById(id: number): Observable<ProjectResponse> {
    return this.http.get<ProjectResponse>(`${this.apiUrl}/${id}`);
  }

  getByUserId(userId: number): Observable<ProjectResponse[]> {
    return this.http.get<ProjectResponse[]>(`${this.apiUrl}/user/${userId}`);
  }

  add(request: CreateProjectRequest): Observable<void> {
    return this.http.post<void>(this.apiUrl, request);
  }

  update(id: number, request: CreateProjectRequest): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/${id}`, request);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
